jdaviz/configs/default/plugins/viewers.py

```diff
@@ -362,3 +392,413 @@ def _ref_or_id(self):
+        spectrum : :class:`glue.core.data.Data`
+            Data object with the spectrum.
+        color : obj
+            Color value for plotting.
+        alpha : float
+            Alpha value for plotting.
+
+        Returns
+        -------
+        result : bool
+            `True` if successful, `False` otherwise.
+        """
+        # If this is the first loaded data, set things up for unit conversion.
+        if len(self.layers) == 0:
+            reset_plot_axes = True
+        else:
+            # Check if the new data flux unit is actually compatible since flux not linked.
+            try:
+                if self.state.y_display_unit not in ['None', None, 'DN']:
+                    uc.to_unit(data, data.find_component_id("flux"), [1, 1],
+                               u.Unit(self.state.y_display_unit))  # Error if incompatible
+            except Exception as err:
+                # Raising exception here introduces a dirty state that messes up next load_data
+                # but not raising exception also causes weird behavior unless we remove the data
+                # completely.
+                self.session.hub.broadcast(SnackbarMessage(
+                    f"Failed to load {data.label}, so removed it: {repr(err)}",
+                    sender=self, color='error'))
+                self.jdaviz_app.data_collection.remove(data)
+                return False
+            reset_plot_axes = False
+
+        # The base class handles the plotting of the main
+        # trace representing the profile itself.
+        result = super().add_data(data, color, alpha, **layer_state)
+
+        if reset_plot_axes:
+            x_units = data.get_component(self.state.x_att.label).units
+
+            y_axis_component = (
+                'flux' if 'flux' in [comp.label for comp in self.state.layers[0].layer.components]
+                else 'data'
+            )
+            y_units = data.get_component(y_axis_component).units
+            with delay_callback(self.state, "x_display_unit", "y_display_unit"):
+                self.state.x_display_unit = x_units if len(x_units) else None
+                self.state.y_display_unit = y_units if len(y_units) else None
+            self.set_plot_axes()
+
+        self._plot_uncertainties()
+
+        self._plot_mask()
+
+        # Set default linewidth on any created spectral subset layers
+        # NOTE: this logic will need updating if we add support for multiple cubes as this assumes
+        # that new data entries (from model fitting or gaussian smooth, etc) will only be spectra
+        # and all subsets affected will be spectral
+        for layer in self.state.layers:
+            if (isinstance(layer.layer, GroupedSubset)
+                    and get_subset_type(layer.layer) == self._default_profile_subset_type
+                    and layer.layer.data.label == data.label):
+                layer.linewidth = 3
+
+        return result
+
+    def _plot_mask(self):
```

Comment:
this all seems to be uncovered code... is it new or just moved?

Reply:
Just moved.